Metadata code refactoring

GernotMaier opened the pull request:
(sorry, this is a difficult one)

This is a major pull request with hopefully improvements on how to handle the metatdata required for each input to a simtool and generate for each output of a simtool. It simplifies the metatdata treatment (note: more lines deleted than created), but unfortunately metadata collection is still quite fine tuned.

The problem with the current implementation is outlined in issue #678 .

Changes:

- use [jsonschema](https://python-jsonschema.readthedocs.io/en/stable/) to define metadata
- encoded in [simtools/schemas/metadata.schema.yml](simtools/schemas/metadata.schema.yml):
   - allows for a full description of all fields (including examples, allowed values, types, etc)
- new application: `simtools-generate-default-metadata-file` to generate an example file from default values
- allows to use the application `simtools-validate-schema-files` to validate metadata
  - good for testing: metadata generated with `simtools-generate-default-metadata-file` must pass `simtools-validate-schema-files`
  - renamed this application to `simtools-validate-file-using-schema` (which is clearer, as we don't validate a schema but use a schema to validate a file)
- using  the jsonschema package allows to remove `simtools/data_model/validate_schema.py`
- changes to pip packaging so that simtools/schemas/*.yml is included in the package
- (a surprising amount of language typos fixed)
- updated all test files in tests/resources/*meta* (the pass now `simtools/applications/validate_file_using_schema.py`

VictorBarbosaMartins commented on simtools/schemas/metadata.schema.yml:
```diff
@@ -0,0 +1,511 @@
+                type: object
+                additionalProperties: false
+                properties:
+                  NAME:
```
I've seen there are some differences when comparing the table in the pdf with the implementation here. For example, in the pdf we only use one level of subitem, but here there are multiple (e.g.Product/model/name, while in the pdf document we have Product/data model name. I'm commenting this because I'm not sure whether we need to be that faithful to the pdf table?

GernotMaier replied:
Did you check A.2.3 in the document? There are several hierarchies. Or do you mean something else?

VictorBarbosaMartins replied:
Ok! I was looking in the table (figure A.2). That is fine then.

VictorBarbosaMartins commented on simtools/schemas/metadata.schema.yml:
Just to make sure I am clear about what I reviewed, I scrolled though this file but I did not cross checked every single information here

GernotMaier replied:
OK, this should be fine.

VictorBarbosaMartins commented on tests/resources/MLTdata-preproduction.meta.yml:
```diff
@@ -1,55 +1,73 @@
+      MODEL:
+        NAME: simpipe-schema
+        VERSION: 0.1.0
+        URL: https://github.com/gammasim/workflows/schema
```
what if this schema changes and the url of an old metadata file points to a new version of the structure? shouldnt we include the version of the workflow repository?

VictorBarbosaMartins commented on tests/unit_tests/data_model/test_metadata_collector.py:
```diff
@@ -240,25 +234,102 @@ def test_collect_schema_dict(args_dict_site, tmp_test_directory):
+    _dict_test_1 = _collector._process_metadata_from_file(meta_dict_1)
+    assert _dict_test_1 == meta_dict_1
+
+    meta_dict_2 = {"cta": {"PRODUCT": {"description": "This is a sample description"}}}
+    _dict_test_1 = _collector._process_metadata_from_file(meta_dict_2)
+
+    meta_dict_3 = {"cta": {"PRODUCT": {"description": "This is a sample\n description"}}}
+    _dict_test_1 = _collector._process_metadata_from_file(meta_dict_3)
```
hmm, it seems copilot is repeating tests? I don't think we need to repeat the test with different inputs

GernotMaier replied:
There is a small difference (and not from copilot): the second statement does include a `\n` and we want to test that it is removed.

VictorBarbosaMartins replied:
Ok!

GernotMaier commented:
Thanks for the review!

The application where actually essential for this pull requests, as they allowed me to test the new implementation.

Everything else is fine and I am going to address your comments now.

> Uff, this was a long review. I know this is a one-time PR with a very nice update on the metadata handling, but let's try to keep PRs small. The applications, for example, could have waited for the next PR.
> 
> In any case, thanks a lot for these updates, as we have spoke early today, this is a major improvement and therefore I also dedicated more time to the review. I provide some comments below. The main issue I raise is in regard to the collection of functions in metadata_model and the lack of an appropriate class there. We can further discuss it on Wednesday if you prefer.
> 
> Please note that I did not double check all entries in the metadata file (should I?). I also did not look very deeply into the unit tests to check whether all functionalities are being tested, though I expect them to.
> 
> Thanks!

GernotMaier commented on docs/source/applications.rst:
```diff
@@ -49,6 +49,12 @@ derive_mirror_rnda
 .. automodule:: derive_mirror_rnda
    :members:
 
+generate_default_metadata_file
```
I don't expect that the general user will need this, but we have a couple of applications like this.  For now, I am not worried about too many applications in the application folder (I would prefer to have many more there :-)). We can start with subdirectories when it gets painful.

GernotMaier commented on pyproject.toml:
```diff
@@ -81,7 +83,7 @@ simtools-tune-psf = "simtools.applications.tune_psf:main"
 simtools-validate-camera-fov = "simtools.applications.validate_camera_fov:main"
 simtools-validate-optics = "simtools.applications.validate_optics:main"
-simtools-validate-schema-files = "simtools.applications.validate_schema_files:main"
+simtools-validate-file-using-schema = "simtools.applications.validate_file_using_schema:main"
```
Good point (and not picky - this improves maintenance.

Fixed.

GernotMaier commented on simtools/applications/generate_default_metadata_file.py:
```diff
@@ -0,0 +1,89 @@
+"""
+    Summary
+    -------
+    Generate a default simtools metadata file from a json schema.
```
Not sure what to add. It generates a metadata file from a json schema.

I don't like adding the current dict the application will return, this might be outdated very soon.

I've added an example.

GernotMaier commented on simtools/applications/generate_default_metadata_file.py:
```diff
@@ -0,0 +1,89 @@
+
+    Command line arguments
+    ----------------------
+    schema (str, optional)
```
No, `default` means to return a dict filled with the default values as given in the schema. I see the confusion, would `generate_metadata_file_with_defaults` be better? Getting a bit long.

VictorBarbosaMartins replied:
Ok! no, that is too long, let's keep it like it is now.

GernotMaier commented on simtools/applications/generate_default_metadata_file.py:
```diff
@@ -0,0 +1,89 @@
+    default_values = metadata_model.top_level_reference_schema(args_dict["schema"])
+
+    if args_dict["output_file"] is None:
+        print(default_values)
```
But this is not an information, but an actual output (the product of this application). So I think print is the right thing.

VictorBarbosaMartins replied:
I see your point, but I expect the output of the application to be the file, as suggested by the name of the application. On the other hand, we have an application called `print_array_elements` that has no print statements in it.

GernotMaier replied:
`print_array_elements` is printing the telescope lists to screen (when no export/output_file option is set).

So my suggestion would be to remove the `_file` and call it `generate_default_metadata`. Would that be fine?

(and note that many of our application names are all over the place and not consistent, this is something to fix for the future)

VictorBarbosaMartins replied:
ok! thank you

GernotMaier commented on simtools/applications/generate_default_metadata_file.py:
```diff
@@ -0,0 +1,89 @@
+        required=False,
+    )
+    config.parser.add_argument(
+        "--output_file",
```
I used it a lot with just printing it to the screen, so I think it should be not required.

Good suggestion of adding a sentence that if output_file is not given, then the dict is printed to the screen. I've added this.

GernotMaier commented on simtools/applications/validate_file_using_schema.py:
```diff
@@ -0,0 +1,77 @@
+"""
+    Summary
+    -------
+    Validate file (e.g., metadata) using a json schema file.
```
Hmm, again not sure what would help.

- json schema is (almost?) an industry standard
- there are different places where they are located, and they might also change in future. 

This tool works with any json schema, so my feeling is that this is sufficient.

GernotMaier replied:
I am really not convinced and I actually think many of our docstrings are too long and duplicated information. I would prefer to keep it short.

GernotMaier commented on simtools/applications/validate_file_using_schema.py:
```diff
@@ -0,0 +1,77 @@
+
+        simtools-validate-file-using-schema \
+         --file_name tests/resources/MST_mirror_2f_measurements.schema.yml \
+         --schema jsonschema.yml
```
Good point - I've added an example with files from the schema / testing area.

GernotMaier commented on simtools/applications/validate_file_using_schema.py:
```diff
@@ -0,0 +1,77 @@
+    Command line arguments
+    ----------------------
+    file_name (str)
+      input file to be validated
```
Not sure if this is more information. A file is a data product, so this would not work. Why does 'input file to be validated' is not sufficient, doesn't it say everything?

VictorBarbosaMartins replied:
input file can be everything. a histogram? we can mention it has to be either a metadata or a data product. Wouldnt that be better?

GernotMaier replied:
Added that it needs to be a yaml file.

GernotMaier commented on simtools/applications/validate_file_using_schema.py:
```diff
@@ -0,0 +1,77 @@
+    try:
+        with open(args_dict["file_name"], "r", encoding="utf-8") as file:
+            data = yaml.safe_load(file)
+    except FileNotFoundError:
```
Done.

GernotMaier commented on simtools/data_model/metadata_collector.py:
```diff
@@ -15,7 +15,7 @@ class MetadataCollector:
     Collects and combines metadata associated with the current activity
     (e.g., the execution of an application).
-    Follows CTAO top-level metadata definition.
+    Depends on and fine tuned to CTAO top-level metadata definition.
```
Those links have a short life time and I consciously did not add a link. 

I have tried and minimize the dependency on the actual data model, but unfortunately couldn't find a way to remove it.

GernotMaier commented on simtools/data_model/metadata_collector.py:
```diff
@@ -112,22 +108,21 @@ def _fill_top_level_meta_from_file(self, top_level_dict):
+            _input_meta = gen.collect_data_from_yaml_or_dict(
+                in_yaml=self.args_dict.get("input_meta", None), in_dict=None
+            )
+        except gen.InvalidConfigData:
```
The last statement of `collect_data_from_or_dict` is a raise, and I am using the function as it is. I prefer try..except statements to ifs, think this improves readability. Is this ok with you?

VictorBarbosaMartins replied:
The last statement is a raise but it only gets there if you don't use the flag `allow_empty=True`. My point is that if there is an error when opening the file `collect_data_from_yaml_or_dict` will handle it. the try and expect is not necessary in my opinion.

GernotMaier replied:
So help me, what is wrong:

We are using the default `allow_empty=False` setting. If `input_meta` is not set on the command line, we are calling `collect_data_from_yaml_or_dict(None, None, None)`. This should raise an exception, as we cannot continue working with an empty metadata dict.

VictorBarbosaMartins replied:
Ok, we can leave it like it is. What the "try and except" is doing is basically changing the error message.

GernotMaier commented on simtools/data_model/metadata_collector.py:
```diff
@@ -112,22 +108,21 @@ def _fill_top_level_meta_from_file(self, top_level_dict):
+                in_yaml=self.args_dict.get("input_meta", None), in_dict=None
+            )
+        except gen.InvalidConfigData:
+            self._logger.debug("Failed reading metadata from file.")
```
Yes! Fixed.

GernotMaier commented on simtools/data_model/metadata_collector.py:
```diff
@@ -112,22 +108,21 @@ def _fill_top_level_meta_from_file(self, top_level_dict):
+            self._logger.debug("Failed reading metadata from file.")
+            raise
+
+        metadata_model.validate_schema(_input_meta, None)
```
The default schema file is part of simtools and located in `simtools/schema/metadata.schema.yml`. This is a fixed model of the simtools metadata and should not change. You can see in `metadata_model:load_schema` this file is loaded for the case of `schema_file is None`.

VictorBarbosaMartins replied:
ok, thank you! it makes sense then. I thought this could be bypassing the validation but that doesnt seem to be the case.

GernotMaier commented on simtools/data_model/metadata_collector.py:
```diff
@@ -309,3 +304,72 @@ def _fill_context_sim_list(product_list, new_entry_dict):
+    def _process_metadata_from_file(self, meta_dict):
+        """
+        Process metadata from file to ensure compatibility
+        with metadata model.
```
Thanks.!

GernotMaier commented on simtools/data_model/metadata_collector.py:
```diff
@@ -309,3 +304,72 @@ def _fill_context_sim_list(product_list, new_entry_dict):
+
+        meta_dict = gen.change_dict_keys_case(meta_dict, True)
+        try:
+            meta_dict["cta"]["product"]["description"] = self._remove_line_feed(
```
I think the way it is improves readability. If you browse through `_process_metadata_from_file`, it is hopefully clear what the two function call are doing. Moving the code into the function would require to add comments, which is not so nice.

GernotMaier commented on simtools/data_model/metadata_collector.py:
```diff
@@ -309,3 +304,72 @@ def _fill_context_sim_list(product_list, new_entry_dict):
+
+    def _copy_metadata_context_lists(self, top_level_dict, _input_meta, key):
+        """
+        Copy list-type metadata from file.
```
I understand your comment about the function name and renamed it to `copy_list_type_metadata`, which hopefully makes it clearer.

Note that for now this deals only with the context lists, but it possibly (likely) will be used for other lists in future (requiring to add the top-level key (i.e., `context in this case) to a function argument.

GernotMaier commented on simtools/data_model/metadata_model.py:
```diff
@@ -1,281 +1,178 @@
-* default: default value
-
-These definitions are used for:
+Follows CTAO top-level data model definition.
```
see comment earlier - links have a very short lifetime and I would prefer not.

GernotMaier commented on simtools/data_model/metadata_model.py:
```diff
@@ -1,281 +1,178 @@
-def top_level_reference_schema():
+def top_level_reference_schema(schema_file=None):
     """
     Reference schema following the CTA Top-Level Data Model.
```
See reply above.

GernotMaier commented on simtools/data_model/metadata_model.py:
```diff
@@ -1,281 +1,178 @@
-
-
-def top_level_reference_schema():
+def top_level_reference_schema(schema_file=None):
```
Good point, that is a leftover from the old implementation where the function was doing different things.

I've named it now `get_default_metadata_dict()`, which is what it actually does.

GernotMaier commented on simtools/data_model/metadata_model.py:
```diff
@@ -1,281 +1,178 @@
     """
-    Reference model data for input metata data of type documentlist or instrumentlist
+    Resolve references in yaml data and expand the
+    received dictionary accordingly.
```
Thanks!

GernotMaier commented on tests/integration_tests/test_applications.py:
```diff
@@ -709,8 +709,8 @@
-    # validate_schema_files
-    "validate_schema_files": [
+    # validate_file_using_schema
+    "validate_file_using_schema": [
```
Not sure what you mean, `validate_file_using_schema` doesn't write anything to an output file. All what it does is checks, and prints and ok (or error) to the screen.

GernotMaier commented on simtools/data_model/metadata_model.py:
```diff
@@ -1,281 +1,178 @@
 
-    msg = f"Invalid schema list: {schema_list}"
-    raise InvalidSchemaList(msg)
+    def expand_ref(ref):
```
This is a 'sub function' (not sure what it is called in python) - a function inside a function. This should be covered by `test_resolve_references`.

VictorBarbosaMartins replied:
Hmm, I am not sure that is a good practice. The sub function might be used in a second function in the future or it could even be used independently, in which case we miss the test. If we want to speed up the test, we could use a mocker, but in this case the time to run the test is not an issue.

GernotMaier replied:
This is standard python practice and we can find it a couple of times in simtools.

GernotMaier commented on simtools/data_model/metadata_model.py:
```diff
@@ -1,281 +1,178 @@
+    return resolve_references_recursive(yaml_data)
+
+
+def fill_defaults(schema):
```
But it fills something: it takes the schema, extracts all the defaults (which requires some gymnastics as you see) and fill a new dict with those defaults.

VictorBarbosaMartins replied:
ok! then I would prefer to name it `fill_dict_with_defaults` for example, indicating what is being filled (since it is not a class instance)

GernotMaier commented on simtools/data_model/metadata_model.py:
```diff
@@ -1,281 +1,178 @@
+    schema = resolve_references(schema)
+
+    def fill_defaults_recursive(subschema, current_dict):
+        if "properties" in subschema:
```
Good point. I've added a try statement (and also the corresponding lines to the unit tests).

GernotMaier commented:
@VictorBarbosaMartins - thanks again for the review, a couple of good points which improved the code.

I think I have addressed all your points, let me know if not.

VictorBarbosaMartins commented on simtools/data_model/metadata_model.py:
```diff
@@ -1,281 +1,178 @@
-    Reference data model scheme for input metadata.
-    Describes metadata provided for input to simtools applications.
+    schema = load_schema(schema_file)
+    return fill_defaults(schema["definitions"]["CTA"])
```
I understand. If it is your choice I am fine with that. Maybe in the future, in case the the module gets more complicated we think about changing. Thanks for the explanation!

VictorBarbosaMartins commented:
> a50b9a5

Hi Gernot. Thank you for considering the comments. I hope they help to improve the code. I left some open items but in general it is already very good! Let me know when you are done with the new changes and I will approve.

GernotMaier commented:
@VictorBarbosaMartins - thanks for the detailed review. I think everything is addressed now. Let me know if I can merge (after the tests passed).

VictorBarbosaMartins reviewed (approved):
I am fine with the changes, so I am approving. Thanks for the implementation and for taking my comments into consideration. It was really helpful to sit together before the review, let's try to do that more often for more complicated PRs.
